README.md
# Image Resizer And Positon Watermark for Cordova #
By: Protonet GmbH

Authors: Joschka Schulz

## Adding the Plugin ##

Use the Cordova CLI and type in the following command:

```
// This plugin uses the cordova-plugin-camera
cordova plugin add cordova-plugin-camera

// This plugin
cordova plugin add https://github.com/azhengyongqin/cordova-plugin-image-resizer-watermark
```

## Sample Code

At the moment the plugin is available on android, iOS and windows

### resize

    window.ImageResizer.resize(options, success, failed);
    
### Options
  - **position**(String):The position of geographic location information
  - **longitude**(String)
  - **latitude**(String)
  - **uri**(String): The Uri for the image on the device to get scaled
  - **folderName**(String, required on Android): The name of the folder the image should be put in **android only**
  - **fileName**(String, required on iOS): A custom name for the file. Default name is a timestamp.
  - **quality**(Number): Quality given as Number for the quality of the new image **android and iOS only**
  - **width**(Number): The width of the new image,
  - **height**(Number): The height of the new image
  - **base64**(Boolean): Whether or not to return a base64 encoded image string instead of the path to the resized image **iOS only**

### Android Example
    var options = {
          uri: uri,
          folderName: "Protonet Messenger",
          quality: 90,
          width: 1280,
          height: 1280,
          base64: true};

    window.ImageResizer.resize(options,
      function(image) {
         // success: image is the new resized image
      }, function() {
        // failed: grumpy cat likes this function
      });


CHANGELOG.md
# version 0.2.0

- Oprion to rotate images based on EXIF data, and Base64 implementation for Android [PR #43](https://github.com/JoschkaSchulz/cordova-plugin-image-resizer/pull/43)

- change name to `cordova-plugin-simple-image-resizer` because `cordova-plugin-image-resizer` was already taken on npm

known issues:

- Build failing (Unknown property `ANDROID_EXIFINTERFACES_VERSION`) [Issue #46](https://github.com/JoschkaSchulz/cordova-plugin-image-resizer/issues/46)


src/ios/ImageResizer.m
#import "ImageResizer.h"
#import <Cordova/CDV.h>
#import <Cordova/CDVPluginResult.h>
#import <AssetsLibrary/AssetsLibrary.h>

#define PROTONET_PHOTO_PREFIX @"protonet_"
#define TIMESTAMP [NSString stringWithFormat:@"%f",[[NSDate date] timeIntervalSince1970] * 1000]

@implementation ImageResizer {
    UIImage* sourceImage;
}

- (void) resize:(CDVInvokedUrlCommand*)command
{
    __block PHImageRequestOptions * imageRequestOptions = [[PHImageRequestOptions alloc] init];
    
    imageRequestOptions.synchronous = YES;
    
    NSLog(@"IMAGE RESIZER START ----------------------------------------------------------------------------");
    
    // get the arguments and the stuff inside of it
    NSDictionary* arguments = [command.arguments objectAtIndex:0];
    NSString* imageUrlString = [arguments objectForKey:@"uri"];
    NSLog(@"Image Resizer Image URL : %@",imageUrlString);
    
    NSString* quality = [arguments objectForKey:@"quality"];
    CGSize frameSize = CGSizeMake([[arguments objectForKey:@"width"] floatValue], [[arguments objectForKey:@"height"] floatValue]);
    NSString* fileName = [arguments objectForKey:@"fileName"];
    
    BOOL asBase64 = [[arguments objectForKey:@"base64"] boolValue];
    BOOL fixRotation = [[arguments objectForKey:@"fixRotation"] boolValue];
    
    //获取经纬度
    self->longitude = [arguments objectForKey:@"longitude"];
    self->latitude = [arguments objectForKey:@"latitude"];
    self->locationDescribe = [arguments objectForKey:@"position"];
    
    // Check if the file is a local file, and if so, read with file manager to avoid NSUrl -1022 error
    if ([[NSFileManager defaultManager] fileExistsAtPath:imageUrlString]){
        sourceImage = [UIImage imageWithData: [[NSFileManager defaultManager] contentsAtPath:imageUrlString]];
    }else {
        sourceImage = [UIImage imageWithData: [NSData dataWithContentsOfURL: [NSURL URLWithString:imageUrlString]]];
    }
    
    int rotation = 0;
    
    switch ([sourceImage imageOrientation]) {
        case UIImageOrientationUp:
            rotation = 0;
            break;
        case UIImageOrientationDown:
            rotation = 180;
            break;
        case UIImageOrientationLeft:
            rotation = 270;
            break;
        case UIImageOrientationRight:
            rotation = 90;
            break;
        default:
            break;
    }
    
    PHFetchResult *savedAssets = [PHAsset fetchAssetsWithLocalIdentifiers:@[fileName] options:nil];
    [savedAssets enumerateObjectsUsingBlock:^(PHAsset *asset, NSUInteger idx, BOOL *stop) {
        //this gets called for every asset from its localIdentifier you saved
        
        [[PHImageManager defaultManager]
         requestImageDataForAsset:asset
         options:imageRequestOptions
         resultHandler:^(NSData *imageData, NSString *dataUTI,
                         UIImageOrientation orientation,
                         NSDictionary *info)
         {
             sourceImage  = [UIImage imageWithData:imageData];
         }];
        
    }];
    
    NSLog(@"image resizer:%@",  (sourceImage ? @"image exists" : @"null" ));
    
    UIImage *tempImage = nil;
    CGSize targetSize = frameSize;
    
    CGRect thumbnailRect = CGRectMake(0, 0, 0, 0);
    thumbnailRect.origin = CGPointMake(0.0,0.0);
    
    // get original image dimensions
    CGFloat heightInPoints = sourceImage.size.height;
    CGFloat heightInPixels = heightInPoints * sourceImage.scale;
    CGFloat widthInPoints = sourceImage.size.width;
    CGFloat widthInPixels = widthInPoints * sourceImage.scale;
    
    // calculate the target dimensions in a way that preserves the original aspect ratio
    CGFloat newWidth = targetSize.width;
    CGFloat newHeight = targetSize.height;
    
    if (heightInPixels > widthInPixels) {
        // vertical image: use targetSize.height as reference for scaling
        newWidth = widthInPixels * newHeight / heightInPixels;
    } else {
        // horizontal image: use targetSize.width as reference
        newHeight = heightInPixels * newWidth / widthInPixels;
    }
    
    thumbnailRect.size.width  = newWidth;
    thumbnailRect.size.height = newHeight;
    targetSize.width = newWidth;
    targetSize.height = newHeight;
    
    UIGraphicsBeginImageContext(targetSize);
    [sourceImage drawInRect:thumbnailRect];
    
    tempImage = UIGraphicsGetImageFromCurrentImageContext();
    NSLog(@"image resizer:%@",  (tempImage  ? @"image exsist" : @"null" ));
    
    if(fixRotation){
        tempImage = [self rotateImage:tempImage withRotation:rotation];
    }
    
    UIGraphicsEndImageContext();
    NSData *imageData = UIImageJPEGRepresentation(tempImage, [quality floatValue] / 100.0f );
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString *cachesDirectory = [paths objectAtIndex:0];
    BOOL isDir = NO;
    NSError *error;
    if (! [[NSFileManager defaultManager] fileExistsAtPath:cachesDirectory isDirectory:&isDir] && isDir == NO) {
        [[NSFileManager defaultManager] createDirectoryAtPath:cachesDirectory withIntermediateDirectories:NO attributes:nil error:&error];
    }
    NSString *imagePath =[cachesDirectory stringByAppendingPathComponent:[NSString stringWithFormat:@"img%@.jpeg", TIMESTAMP]];
    CDVPluginResult* result = nil;
    
    if (asBase64) {
        NSData *imageBase64 = [imageData base64EncodedDataWithOptions:NSDataBase64Encoding64CharacterLineLength];
        NSString *imageBase64String = [[NSString alloc] initWithData:imageBase64 encoding:NSUTF8StringEncoding];
        NSString *imageBase64URL = [NSString stringWithFormat:@"%@%@", @"data:image/jpeg;base64,", imageBase64String];
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:imageBase64URL];
    }
    else if (![imageData writeToFile:imagePath atomically:NO])
    {
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_IO_EXCEPTION messageAsString:@"error save image"];
    }
    else
    {
        result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[[NSURL fileURLWithPath:imagePath] absoluteString]];
    }
    
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (UIImage*) rotateImage:(UIImage*) image withRotation:(int) rotation{
    CGFloat rot = rotation * M_PI / 180;
    
    // Calculate Destination Size
    CGAffineTransform t = CGAffineTransformMakeRotation(rot);
    CGRect sizeRect = (CGRect) {.size = image.size};
    CGRect destRect = CGRectApplyAffineTransform(sizeRect, t);
    CGSize destinationSize = destRect.size;
    
    // Draw image
    UIGraphicsBeginImageContext(destinationSize);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextTranslateCTM(context, destinationSize.width / 2.0f, destinationSize.height / 2.0f);
    CGContextRotateCTM(context, rot);
    [image drawInRect:CGRectMake(-image.size.width / 2.0f, -image.size.height / 2.0f, image.size.width, image.size.height)];
    
    // Save image
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return [self addTextWatermark:newImage];
}


//添加水印
-(UIImage *)addTextWatermark:(UIImage *)img
{
    NSString* latitudeAndLongitude = [NSString stringWithFormat:@"经纬度：(%@，%@)", self->longitude,self->latitude];
    NSLog(@"添加水印：%@",latitudeAndLongitude);
    NSString* time = [ImageResizer getCurrentTimes];
    NSLog(@"添加水印：%@",time);
    int w = img.size.width;
    int h = img.size.height;
    //UIGraphicsBeginImageContext创建一个基于位图的上下文(context),并将其设置为当前上下文(context)
    UIGraphicsBeginImageContext(img.size);
    
    [img drawInRect:CGRectMake(0, 0, w, h)];
    
    NSDictionary *localAttr = @{NSFontAttributeName: [UIFont boldSystemFontOfSize:40],   //设置字体
                                NSForegroundColorAttributeName : [UIColor whiteColor]      //设置字体颜色
                                };
    NSDictionary *llAttr = @{NSFontAttributeName: [UIFont boldSystemFontOfSize:45],   //设置字体
                             NSForegroundColorAttributeName : [UIColor whiteColor]      //设置字体颜色
                             };
    NSDictionary *timeAttr = @{NSFontAttributeName: [UIFont boldSystemFontOfSize:60],   //设置字体
                               NSForegroundColorAttributeName : [UIColor whiteColor]      //设置字体颜色
                               };
    
    
    [self->locationDescribe drawInRect:CGRectMake(40, h - 40 - 40, w, 70) withAttributes:localAttr];        //左下角
    [latitudeAndLongitude drawInRect:CGRectMake(40, h - 45 - 50*2, w, 70) withAttributes:llAttr];        //左下角
    [time drawInRect:CGRectMake(40, h - 70 - 50*3, w, 70) withAttributes:timeAttr];        //左下角
    UIImage *newImg = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newImg;
}
//获取当前的时间
+(NSString*)getCurrentTimes{
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    // ----------设置你想要的格式,hh与HH的区别:分别表示12小时制,24小时制
    [formatter setDateFormat:@"时间：YYYY-MM-dd HH:mm:ss"];
    //现在时间,你可以输出来看下是什么格式
    NSDate *datenow = [NSDate date];
    //----------将nsdate按formatter格式转成nsstring
    NSString *currentTimeString = [formatter stringFromDate:datenow];
    NSLog(@"currentTimeString =  %@",currentTimeString);
    return currentTimeString;
}

@end


src/android/ImageResizer.java
package info.protonet.imageresizer;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.media.ExifInterface;
import android.net.Uri;
import android.os.Environment;
import android.util.Base64;
import android.util.Log;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.apache.cordova.camera.FileHelper;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class ImageResizer extends CordovaPlugin {
  private static final int ARGUMENT_NUMBER = 1;
  public CallbackContext callbackContext;

  private String uri;
  private String folderName;
  private String fileName;

  /**
   * 经度
   */
  private String longitude;
  /**
   * 纬度
   */
  private String latitude;
  /**
   * 位置
   */
  private String position;
  private int quality;
  private int width;
  private int height;
  private boolean fixRotation = false;
  private boolean base64 = false;

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    try {
      this.callbackContext = callbackContext;

      if (action.equals("resize")) {
        checkParameters(args);

        // get the arguments
        JSONObject jsonObject = args.getJSONObject(0);
        uri = jsonObject.getString("uri");
        folderName = null;
        if (jsonObject.has("folderName")) {
          folderName = jsonObject.getString("folderName");
        }
        fileName = null;
        if (jsonObject.has("fileName")) {
          fileName = jsonObject.getString("fileName");
        }
        longitude = null;
        if (jsonObject.has("longitude")) {
          longitude = jsonObject.getString("longitude");
        }
        latitude = null;
        if (jsonObject.has("latitude")) {
          latitude = jsonObject.getString("latitude");
        }
        position = null;
        if (jsonObject.has("position")) {
          position = jsonObject.getString("position");
        }


        quality = jsonObject.getInt("quality");
        width = jsonObject.getInt("width");
        height = jsonObject.getInt("height");
        if (jsonObject.has("fixRotation")) {
          fixRotation = jsonObject.getBoolean("fixRotation");
        }
        if (jsonObject.has("base64")) {
          base64 = jsonObject.getBoolean("base64");
        }

        // load the image from uri
        Bitmap bitmap = loadScaledBitmapFromUri(uri, width, height);

        if (bitmap == null) {
          Log.e("Protonet", "There was an error reading the image");
          callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR));
          return false;
        }

        if (fixRotation) {
          // Get the exif rotation in degrees, create a transformation matrix, and rotate
          // the bitmap
          int rotation = getRoationDegrees(getRotation(uri));
          Matrix matrix = new Matrix();
          if (rotation != 0f) {
            matrix.preRotate(rotation);
          }
          bitmap = Bitmap.createBitmap(
            bitmap,
            0,
            0,
            bitmap.getWidth(),
            bitmap.getHeight(),
            matrix,
            true);
        }

        if (base64) {
          // convert the bitmap to a b64 string and return
          ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
          bitmap.compress(CompressFormat.JPEG, 100, byteArrayOutputStream);
          byte[] byteArray = byteArrayOutputStream.toByteArray();
          String encoded = Base64.encodeToString(byteArray, Base64.DEFAULT);
          callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK,
            "data:image/jpeg;base64," + encoded));
        } else {
          // save the image as jpeg on the device
          Uri scaledFile = saveFile(bitmap);

          if (scaledFile == null) {
            Log.e("Protonet", "There was an error saving the thumbnail");
            callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR));
            return false;
          }
          callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK,
            scaledFile.toString()));
        }
        return true;
      } else {
        callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR));
        return false;
      }
    } catch (JSONException e) {
      Log.e("Protonet", "JSON Exception during the Image Resizer Plugin... :(");
    }
    callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR));
    return false;
  }

  /**
   * Gets the image rotation from the image EXIF Data
   *
   * @param exifOrientation ExifInterface.ORIENTATION_* representation of the rotation
   * @return the rotation in degrees
   */
  private int getRoationDegrees(int exifOrientation) {
    if (exifOrientation == ExifInterface.ORIENTATION_ROTATE_90) {
      return 90;
    } else if (exifOrientation == ExifInterface.ORIENTATION_ROTATE_180) {
      return 180;
    } else if (exifOrientation == ExifInterface.ORIENTATION_ROTATE_270) {
      return 270;
    }
    return 0;
  }

  /**
   * Gets the image rotation from the image EXIF Data
   *
   * @param uriString the URI of the image to get the rotation for
   * @return ExifInterface.ORIENTATION_* representation of the rotation
   */
  private int getRotation(String uriString) {
    try {
      ExifInterface exif = new ExifInterface(uriString);
      return exif.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL);
    } catch (IOException e) {
      return ExifInterface.ORIENTATION_NORMAL;
    }
  }

  /**
   * Loads a Bitmap of the given android uri path
   *
   * @params uri the URI who points to the image
   **/
  private Bitmap loadScaledBitmapFromUri(String uriString, int width, int height) {
    try {
      BitmapFactory.Options options = new BitmapFactory.Options();
      options.inJustDecodeBounds = true;
      BitmapFactory.decodeStream(FileHelper.getInputStreamFromUriString(uriString, cordova), null, options);

      //calc aspect ratio
      int[] retval = calculateAspectRatio(options.outWidth, options.outHeight);

      options.inJustDecodeBounds = false;
      options.inSampleSize = calculateSampleSize(options.outWidth, options.outHeight, width, height);
      Bitmap unscaledBitmap = BitmapFactory.decodeStream(FileHelper.getInputStreamFromUriString(uriString, cordova), null, options);
      if (unscaledBitmap != null) {
        return Bitmap.createScaledBitmap(unscaledBitmap, retval[0], retval[1], true);
      }
    } catch (FileNotFoundException e) {
      Log.e("Protonet", "File not found. :(");
    } catch (IOException e) {
      Log.e("Protonet", "IO Exception :(");
    } catch (Exception e) {
      Log.e("Protonet", e.toString());
    }
    return null;
  }

  private Uri saveFile(Bitmap bitmap) {
    @SuppressLint("SimpleDateFormat")
    SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
    Date date = new Date();
    String dateStr = "时间：" + dateFormat.format(date);
    String latitudeAndLongitudeStr = "";
    if (!"".equals(longitude) && !"".equals(latitude)) {
      latitudeAndLongitudeStr = "经纬度：（" + longitude + "，" + latitude + ")";
    }


    //添加水印
    bitmap = addTextWatermark(bitmap, dateStr, position, latitudeAndLongitudeStr);

    File folder = null;
    if (folderName == null) {
      folder = new File(this.getTempDirectoryPath());
    } else {
//      if (folderName.contains("/")) {
//        folder = new File(folderName.replace("file://", ""));
//      } else {
      folder = new File(Environment.getExternalStorageDirectory() + "/" + folderName);
//      }
    }
    boolean success = true;
    if (!folder.exists()) {
      success = folder.mkdirs();
    }

    if (success) {
      if (fileName == null) {
        fileName = System.currentTimeMillis() + ".jpg";
      }
      File file = new File(folder, fileName);
      if (file.exists()) file.delete();
      try {
        FileOutputStream out = new FileOutputStream(file);
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, out);
        out.flush();
        out.close();
      } catch (Exception e) {
        Log.e("Protonet", e.toString());
      }
      return Uri.fromFile(file);
    }
    return null;
  }

  /**
   * Figure out what ratio we can load our image into memory at while still being bigger than
   * our desired width and height
   *
   * @param srcWidth
   * @param srcHeight
   * @param dstWidth
   * @param dstHeight
   * @return
   */
  private int calculateSampleSize(int srcWidth, int srcHeight, int dstWidth, int dstHeight) {
    final float srcAspect = (float) srcWidth / (float) srcHeight;
    final float dstAspect = (float) dstWidth / (float) dstHeight;

    if (srcAspect > dstAspect) {
      return srcWidth / dstWidth;
    } else {
      return srcHeight / dstHeight;
    }
  }

  /**
   * Maintain the aspect ratio so the resulting image does not look smooshed
   *
   * @param origWidth
   * @param origHeight
   * @return
   */
  private int[] calculateAspectRatio(int origWidth, int origHeight) {
    int newWidth = width;
    int newHeight = height;

    // If no new width or height were specified return the original bitmap
    if (newWidth <= 0 && newHeight <= 0) {
      newWidth = origWidth;
      newHeight = origHeight;
    }
    // Only the width was specified
    else if (newWidth > 0 && newHeight <= 0) {
      newHeight = (newWidth * origHeight) / origWidth;
    }
    // only the height was specified
    else if (newWidth <= 0 && newHeight > 0) {
      newWidth = (newHeight * origWidth) / origHeight;
    }
    // If the user specified both a positive width and height
    // (potentially different aspect ratio) then the width or height is
    // scaled so that the image fits while maintaining aspect ratio.
    // Alternatively, the specified width and height could have been
    // kept and Bitmap.SCALE_TO_FIT specified when scaling, but this
    // would result in whitespace in the new image.
    else {
      double newRatio = newWidth / (double) newHeight;
      double origRatio = origWidth / (double) origHeight;

      if (origRatio > newRatio) {
        newHeight = (newWidth * origHeight) / origWidth;
      } else if (origRatio < newRatio) {
        newWidth = (newHeight * origWidth) / origHeight;
      }
    }

    int[] retval = new int[2];
    retval[0] = newWidth;
    retval[1] = newHeight;
    return retval;
  }

  private void checkParameters(JSONArray args) {
    if (args.length() != ARGUMENT_NUMBER) {
      callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.INVALID_ACTION));
    }
  }

  private String getTempDirectoryPath() {
    File cache = null;

    // SD Card Mounted
    if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
      cache = new File(Environment.getExternalStorageDirectory().getAbsolutePath() +
        "/Android/data/" + cordova.getActivity().getPackageName() + "/cache/");
    } else {
      // Use internal storage
      cache = cordova.getActivity().getCacheDir();
    }

    // Create the cache directory if it doesn't exist
    cache.mkdirs();
    return cache.getAbsolutePath();
  }


  /**
   * 给一张Bitmap添加水印文字。
   *
   * @param bitmap   源图片
   * @param dataStr  水印文本
   * @param position 位置
   * @return 已经添加水印后的Bitmap。
   */
  private static Bitmap addTextWatermark(Bitmap bitmap, String dataStr, String position, String latitudeAndLongitudeStr) {
    if (position == null) {
      position = "";
    }
    if (isEmptyBitmap(bitmap) || dataStr == null) {
      return null;
    }
    float x = 40.0f;
    float y = bitmap.getHeight() - 160;

    Bitmap ret = bitmap.copy(bitmap.getConfig(), true);
    Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    Canvas canvas = new Canvas(ret);
    paint.setColor(Color.WHITE);
    Rect bounds = new Rect();

    //绘制时间
    paint.setTextSize(50.0f);
    paint.getTextBounds(dataStr, 0, dataStr.length(), bounds);
    y = bitmap.getHeight() - 120;
    canvas.drawText(dataStr, x, y, paint);

    //绘制经纬度
    paint.setTextSize(40.0f);
    paint.getTextBounds(latitudeAndLongitudeStr, 0, latitudeAndLongitudeStr.length(), bounds);
    y = bitmap.getHeight() - 65;
    canvas.drawText(latitudeAndLongitudeStr, x, y, paint);


    //绘制地址描述
    paint.setTextSize(30.0f);
    paint.getTextBounds(position, 0, position.length(), bounds);
    y = bitmap.getHeight() - 20;
    canvas.drawText(position, x, y, paint);
    bitmap.recycle();
    return ret;
  }

  /**
   * Bitmap对象是否为空。
   */
  private static boolean isEmptyBitmap(Bitmap src) {
    return src == null || src.getWidth() == 0 || src.getHeight() == 0;
  }

}
